File: Visual Studio/Fun/The Matrix Text Rain/Core/Backend.cs
using System;
using System.Collections.Generic;

namespace Core
{
    internal class Backend
    {
        private readonly List<List<TheMatrixRaindrop>> rainColumns = new List<List<TheMatrixRaindrop>>();
        private readonly Random random = new Random();
        private readonly double λMutation;
        private readonly double λGenerate;
        private readonly double minimalSpeed;
        private readonly double maximalSpeed;
        private readonly int minimalRaindropSize;
        private readonly int maximalRaindropSize;
        private readonly string characterCandidates;
        private int rows = -1;
        private double lastViewTime;
        private readonly IDictionary<int, Stack<TheMatrixRaindrop>> raindropRecycleBin = new Dictionary<int, Stack<TheMatrixRaindrop>>();
        private readonly Stack<TheMatrixRaindrop> sharedRemoveList = new Stack<TheMatrixRaindrop>();

        public Backend(double λMutation = 0.2, double λGenerate = 1.0, double minimalSpeed = 16.0, double maximalSpeed = 32.0, int minimalRaindropSize = 12, int maximalRaindropSize = 36, string characterCandidates = "0123456789")
        {
            this.λMutation = λMutation;
            this.λGenerate = λGenerate;
            this.minimalSpeed = minimalSpeed;
            this.maximalSpeed = maximalSpeed;
            this.minimalRaindropSize = minimalRaindropSize;
            this.maximalRaindropSize = maximalRaindropSize;
            this.characterCandidates = characterCandidates;
        }

        private char GetRandomCharacter()
        {
            return characterCandidates[random.Next(characterCandidates.Length)];
        }

        private bool ProbabilityGate(double probability)
        {
            return random.NextDouble() < probability;
        }

        private double GetTimeToBirth()
        {
            return Math.Log(1.0 / (1.0 - random.NextDouble()), 2.0) * λGenerate;
        }

        private double GenerateSpeedValue()
        {
            return minimalSpeed + (maximalSpeed - minimalSpeed) * random.NextDouble();
        }

        private IEnumerable<char> GenerateCharacters(int size)
        {
            for (var i = 0; i < size; i++)
            {
                yield return GetRandomCharacter();
            }
        }

        private bool UpdateRaindrop(TheMatrixRaindrop raindrop, double timeElapsed, double mutationProbability)
        {
            var oldIntegerPosition = (int)raindrop.Position;

            raindrop.Position += raindrop.Speed * timeElapsed;

            if (raindrop.Position - raindrop.Size < rows)
            {
                var newIntegerPosition = (int)raindrop.Position;
                var integerStep = newIntegerPosition - oldIntegerPosition;

                // Do character rotation.

                for (var i = raindrop.Size - 1; i >= integerStep; i--)
                {
                    if (ProbabilityGate(mutationProbability))
                    {
                        raindrop.Characters[i] = GetRandomCharacter();
                    }
                    else
                    {
                        raindrop.Characters[i] = raindrop.Characters[i - integerStep];
                    }
                }

                // Fill remaining positions.

                for (var i = 0; i < Math.Min(raindrop.Size, integerStep); i++)
                {
                    raindrop.Characters[i] = GetRandomCharacter();
                }

                return true;
            }
            else
            {
                return false;
            }
        }

        private void RecycleRaindrop(TheMatrixRaindrop raindrop)
        {
            if (raindropRecycleBin.TryGetValue(raindrop.Size, out Stack<TheMatrixRaindrop> bucket))
            {
                bucket.Push(raindrop);
            }
            else
            {
                bucket = new Stack<TheMatrixRaindrop>();

                bucket.Push(raindrop);

                raindropRecycleBin.Add(raindrop.Size, bucket);
            }
        }

        private TheMatrixRaindrop CreateRaindrop(double position, double speed, int size)
        {
            if (raindropRecycleBin.TryGetValue(size, out Stack<TheMatrixRaindrop> bucket) && bucket.Count > 0)
            {
                var result = bucket.Pop();

                result.Reset(position, speed, GenerateCharacters(size));

                return result;
            }
            else
            {
                return new TheMatrixRaindrop(position, speed, GenerateCharacters(size));
            }
        }

        private void UpdateColumn(List<TheMatrixRaindrop> column, double currentTime, double timeElapsed, double mutationProbability)
        {
            foreach (var raindrop in column)
            {
                if (!UpdateRaindrop(raindrop, timeElapsed, mutationProbability))
                {
                    sharedRemoveList.Push(raindrop);
                }
            }

            column.RemoveAll(sharedRemoveList.Contains);

            while (sharedRemoveList.Count > 0)
            {
                RecycleRaindrop(sharedRemoveList.Pop());
            }

            for (var raindropBirthTime = lastViewTime + GetTimeToBirth(); raindropBirthTime <= currentTime; raindropBirthTime += GetTimeToBirth())
            {
                var speed = GenerateSpeedValue();
                var position = speed * (currentTime - raindropBirthTime);
                var size = random.Next(minimalRaindropSize, maximalRaindropSize);

                if (position - size < rows)
                {
                    column.Add(CreateRaindrop(position, speed, size));
                }
            }
        }

        public IReadOnlyList<IReadOnlyList<TheMatrixRaindrop>> GetView(int columns, int rows, double time)
        {
            while (rainColumns.Count < columns)
            {
                rainColumns.Add(new List<TheMatrixRaindrop>());
            }

            while (rainColumns.Count > columns)
            {
                rainColumns[rainColumns.Count - 1].ForEach(RecycleRaindrop);
                rainColumns.RemoveAt(rainColumns.Count - 1);
            }

            this.rows = rows;

            var timeEllapsed = time - lastViewTime;
            var mutationProbability = 1.0 - Math.Pow(2.0, -timeEllapsed / λMutation);

            foreach (var rainColumn in rainColumns)
            {
                UpdateColumn(rainColumn, time, timeEllapsed, mutationProbability);
            }

            lastViewTime = time;

            return rainColumns;
        }
    }
}

File: Rust/matrix-digital-rain/src/backend.rs
use raindrop::*;
use rand::*;
use std::cmp::*;
use utilities::*;

struct Configuration
{
    lambda_mutation: f64,
    lambda_generate: f64,
    minimal_speed: f64,
    maximal_speed: f64,
    minimal_raindrop_size: usize,
    maximal_raindrop_size: usize,
    character_candidates: Vec<char>
}

impl Configuration
{
    fn new() -> Configuration
    {
        return Configuration {
            lambda_mutation: 0.2,
            lambda_generate: 1.0,
            minimal_speed: 16.0,
            maximal_speed: 32.0,
            minimal_raindrop_size: 12,
            maximal_raindrop_size: 36,
            character_candidates: "!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀāĂăĄąĆćĈĉĊċČčĎďĐđĒēĔĕĖėĘęĚěĜĝĞğĠġĢģĤĥĦħĨĩĪīĬĭĮįİıĲĳĴĵĶķĸĹĺĻļĽľĿŀŁłŃńŅņŇňŉŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚśŜŝŞşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſƀƊƏƒƓƠơƯưǂǍǎǏǐǑǒǓǔǕǖǗǘǙǚǛǜǢǣǦǧǪǫǴǵǸǹǺǻǼǽǾǿȘșȚțɃɐɑɒɓɔɕɖɗɘəɚɛɜɞɟɠɡɢɣɤɥɦɧɨɪɫɬɭɮɯɰɱɲɳɴɵɶɸɹɺɻɽɾʀʁʂʃʄʇʈʉʊʋʌʍʎʏʐʑʒʔʕʘʙʜʝʞʟʡʢʤʦʧʰʲʳʷʸʹʻʼʾʿˁˆˇˈˉˊˋˌːˑ˘˙˚˛˝ˠˡˢˣ;΄΅Ά·ΈΉΊΌΎΏΐΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩΪΫάέήίΰαβγδεζηθικλμνξοπρςστυφχψωϊϋόύώϐϑϕϗϙϛϝϡЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏАБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдежзийклмнопрстуфхцчшщъыьэюяѐёђѓєѕіїјљњћќѝўџѢѣѲѳѴѵҐґҒғҖҗҘҙҚқҠҡҢңҪҫҮүҰұҲҳҶҷҺһӀӁӂӏӐӑӔӕӖӗӘәӢӣӦӧӨөӮӯӲӳᵃᵇᵈᵉᵍᵏᵐᵒᵖᵗᵘᵛᶜᶠᶻḆḇḌḍḎḏḐḑḖḗḠḡḤḥḦḧḨḩḪḫḲḳḴḵḶḷḸḹḺḻḾḿṀṁṂṃṄṅṆṇṈṉṒṓṘṙṚṛṜṝṞṟṠṡṢṣṬṭṮṯṾṿẀẁẂẃẄẅẎẏẐẑẒẓẔẕẖẗẞẠạẢảẤấẦầẨẩẪẫẬậẮắẰằẲẳẴẵẶặẸẹẺẻẼẽẾếỀềỂểỄễỆệỈỉỊịỌọỎỏỐốỒồỔổỖỗỘộỚớỜờỞởỠỡỢợỤụỦủỨứỪừỬửỮữỰựỲỳỴỵỶỷỸỹ᾽ι῀῁῍῎῏῝῞῟῭΅`´῾‒–—―‗‘’‚‛“”„†‡•…‰′″‹›‼‽‾ⁿₔ₡₣₤₦₧₩₫€₮₱₲₴₵₸₹₺₽℅ℓ№™Ω℮⅓⅔⅛⅜⅝⅞←↑→↓↔↕↨∂∆∏∑−∕∙√∞∟∩∫≈≠≡≤≥⌂⌐⌠⌡─│┌┐└┘├┤┬┴┼═║╒╓╔╕╖╗╘╙╚╛╜╝╞╟╠╡╢╣╤╥╦╧╨╩╪╫╬□▫▬◊○◌◘◦☺☻☼♀♂♪♫ﬁﬂ".chars().collect(),
        };
    }
}

struct State
{
    configuration: Configuration,
    rows: usize,
    random_number_generator: ThreadRng
}

impl State
{
    fn new() -> State
    {
        return State {
            configuration: Configuration::new(),
            rows: 0,
            random_number_generator: thread_rng()
        };
    }

    fn probability_gate(&mut self, probability: f64) -> bool
    {
        return self.random_number_generator.next_f64() < probability;
    }

    fn get_random_character(&mut self) -> char
    {
        return *self.random_number_generator.choose(self.configuration.character_candidates.as_slice()).unwrap();
    }

    fn get_time_to_birth(&mut self) -> f64
    {
        return ((1.0 / (1.0 - self.random_number_generator.next_f64())) * self.configuration.lambda_generate).log2();
    }

    fn generate_speed_value(&mut self) -> f64
    {
        return self.random_number_generator
                   .gen_range(self.configuration.minimal_speed, self.configuration.maximal_speed);
    }

    fn generate_random_characters(&mut self, size: usize) -> Vec<char>
    {
        return (0..size).map(|_| self.get_random_character()).collect();
    }

    fn update_raindrop(&mut self, raindrop: &mut Raindrop, time_elapsed: f64, mutation_probability: f64) -> bool
    {
        let old_integer_position = raindrop.position as usize;

        raindrop.position += raindrop.speed * time_elapsed;

        if raindrop.position - (raindrop.get_size() as f64) < self.rows as f64
        {
            let new_integer_position = raindrop.position as usize;
            let integer_step = new_integer_position - old_integer_position;

            // Do character rotation.

            for i in (integer_step..raindrop.get_size()).rev()
            {
                if self.probability_gate(mutation_probability)
                {
                    raindrop.characters[i] = self.get_random_character();
                }
                else
                {
                    raindrop.characters[i] = raindrop.characters[i - integer_step];
                }
            }

            // Fill remaining positions.

            for i in 0..min(raindrop.get_size(), integer_step)
            {
                raindrop.characters[i] = self.get_random_character();
            }

            return true;
        }
        else
        {
            return false;
        }
    }

    fn update_column(&mut self, column: &mut Vec<Raindrop>, time_elapsed: f64, mutation_probability: f64)
    {
        let mut remove_list = Vec::new();

        for i in 0..column.len()
        {
            if !self.update_raindrop(&mut column[i], time_elapsed, mutation_probability)
            {
                remove_list.push(i);
            }
        }

        remove_by_indexes(column, &remove_list);

        let mut raindrop_birth_time = self.get_time_to_birth();

        while raindrop_birth_time <= time_elapsed
        {
            let speed = self.generate_speed_value();
            let position = speed * (time_elapsed - raindrop_birth_time);
            let size = self.random_number_generator.gen_range(self.configuration.minimal_raindrop_size,
                                                              self.configuration.maximal_raindrop_size);

            if position - (size as f64) < self.rows as (f64)
            {
                column.push(Raindrop::new(position, speed, self.generate_random_characters(size)));
            }

            raindrop_birth_time += self.get_time_to_birth();
        }
    }
}

pub struct Backend
{
    state: State,
    rain_columns: Vec<Vec<Raindrop>>
}

impl Backend
{
    pub fn new() -> Backend
    {
        return Backend {
            state: State::new(),
            rain_columns: Vec::new()
        };
    }

    pub fn get_view(&mut self, columns: usize, rows: usize, time_ellapsed: f64) -> &[Vec<Raindrop>]
    {
        self.rain_columns.resize(columns, Vec::new());
        self.state.rows = rows;

        let mutation_probability = 1.0 - 2.0f64.powf(-time_ellapsed / self.state.configuration.lambda_mutation);

        for rain_column in &mut self.rain_columns
        {
            self.state.update_column(rain_column, time_ellapsed, mutation_probability);
        }

        return self.rain_columns.as_slice();
    }
}

